content/python_files/imbalance_correction.py

```diff
@@ -0,0 +1,853 @@
+        column_data = {}
+        if hasattr(self, "true_coef"):
+            column_data["Data generating model"] = np.hstack(
+                (self.true_intercept, self.true_coef)
+            )
+        column_data.update(
+            {
+                model_name: np.hstack((model.intercept_, model.coef_.flatten()))
+                for model_name, model in self.models.items()
+            }
+        )
+        pd.DataFrame(
+            column_data,
+            index=np.hstack(["intercept", self.X.columns]),
+        ).plot.barh().set(
+            title="Comparison of the true and learned model parameters",
+            xlabel="Parameter value",
+            ylabel="Parameter name",
+        )
+
+
+population_comparator = ModelComparator(X_future, y_future, context_name="population")
+population_comparator.register_linear_data_generating_model(true_coef, true_intercept)
+population_comparator.register_model("LogReg fit on future data", cheating_model)
+population_comparator.plot_linear_model_parameters()
+
+# %% [markdown]
+#
+# We can see that the learned model parameters nearly recover the data
+# generating parameters. This is expected since the logistic regression model
+# without penalty is well specified for this data generating process and that
+# we have a large number of data points.
+
+# %% [markdown]
+#
+# ## Prevalence shift induced by the data acquisition process
+#
+# The following code simulate what could happen in practice when working in
+# such a setting: we subsample the simulated population to collect all the
+# positive cases and a random sample of negatives from the past data. Mere
+# mortal data scientists cannot `pd.read_csv` from the future, unfortunately.
+#
+# We subsample only the negative cases (control group) to reflect the fact that
+# the negative data is less interesting than the rare positive cases: as such
+# the negative data is in general not archived fully in databases, or even, the
+# features of most of the negative cases are not acquired at all in the first
+# place.
+
+# %%
+X_positive = X_past[y_past == 1].copy()
+y_positive = y_past[y_past == 1].copy()
+
+rng = np.random.default_rng(0)
+negative_indices = rng.choice(
+    np.arange(len(y_past))[y_past == 0],
+    size=3 * len(y_positive),
+    replace=False,
+)
+
+X_observed = pd.concat([X_positive, X_past.iloc[negative_indices]])
+y_observed = pd.concat([y_positive, y_past.iloc[negative_indices]])
+
+X_observed.shape
+# %%
+observed_positive_rate = y_observed.sum() / len(y_observed)
+observed_positive_rate
+
+# %% [markdown]
+#
+# Now that the data is collected, we can train-test split it to reflect what a
+# data scientist would do in practice to train and evaluate their model from
+# the observed data.
+
+# %%
+from sklearn.model_selection import train_test_split
+
+
+X_train, X_test, y_train, y_test = train_test_split(
+    X_observed, y_observed, test_size=0.5, random_state=0
+)
+
+# %%
+logreg_params = dict(penalty=None, tol=1e-8)
```

Comment:
Do we want to use this for the entire notebooks? I see a couple of occurrences where we don't set the tolerance?

Reply:
We could, but initially, I used it to explore the impact of using a non-zero penalty on all the models trained on the small subsampled data to debug other things.